castai/resource_workload_scaling_policy.go

```diff
@@ -0,0 +1,341 @@
+		return diag.FromErr(err)
+	}
+
+	sp := resp.JSON200
+
+	if err := d.Set("name", sp.Name); err != nil {
+		return diag.FromErr(fmt.Errorf("setting name: %w", err))
+	}
+	if err := d.Set("apply_type", sp.ApplyType); err != nil {
+		return diag.FromErr(fmt.Errorf("setting apply type: %w", err))
+	}
+	if err := d.Set("management_option", sp.RecommendationPolicies.ManagementOption); err != nil {
+		return diag.FromErr(fmt.Errorf("setting management option: %w", err))
+	}
+	if err := d.Set("cpu", toResourceMap(sp.RecommendationPolicies.Cpu)); err != nil {
+		return diag.FromErr(fmt.Errorf("setting cpu: %w", err))
+	}
+	if err := d.Set("memory", toResourceMap(sp.RecommendationPolicies.Memory)); err != nil {
+		return diag.FromErr(fmt.Errorf("setting memory: %w", err))
+	}
+
+	return nil
+}
+
+func resourceWorkloadScalingPolicyUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	if !d.HasChanges(
+		"name",
+		"apply_type",
+		"management_option",
+		"cpu",
+		"memory",
+	) {
+		log.Printf("[INFO] scaling policy up to date")
+		return nil
+	}
+
+	client := meta.(*ProviderConfig).api
+	clusterID := d.Get(FieldClusterID).(string)
+	req := sdk.WorkloadOptimizationAPIUpdateWorkloadScalingPolicyJSONBody{
+		Name:      d.Get("name").(string),
+		ApplyType: sdk.WorkloadoptimizationV1ApplyType(d.Get("apply_type").(string)),
+		RecommendationPolicies: sdk.WorkloadoptimizationV1RecommendationPolicies{
+			ManagementOption: sdk.WorkloadoptimizationV1ManagementOption(d.Get("management_option").(string)),
+			Cpu:              toResourcePolicies(d.Get("cpu").([]interface{})[0].(map[string]interface{})),
+			Memory:           toResourcePolicies(d.Get("memory").([]interface{})[0].(map[string]interface{})),
+		},
+	}
+
+	resp, err := client.WorkloadOptimizationAPIUpdateWorkloadScalingPolicyWithResponse(ctx, clusterID, d.Id(), req)
+	if checkErr := sdk.CheckOKResponse(resp, err); checkErr != nil {
+		return diag.FromErr(checkErr)
+	}
+
+	return resourceWorkloadScalingPolicyRead(ctx, d, meta)
+}
+
+func resourceWorkloadScalingPolicyDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*ProviderConfig).api
+	clusterID := d.Get(FieldClusterID).(string)
+
+	resp, err := client.WorkloadOptimizationAPIGetWorkloadScalingPolicyWithResponse(ctx, clusterID, d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	if resp.StatusCode() == http.StatusNotFound {
+		log.Printf("[DEBUG] Scaling policy (%s) not found, skipping delete", d.Id())
+		return nil
+	}
+	if err := sdk.StatusOk(resp); err != nil {
+		return diag.FromErr(err)
+	}
+
+	delResp, err := client.WorkloadOptimizationAPIDeleteWorkloadScalingPolicyWithResponse(ctx, clusterID, d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	if delResp.StatusCode() == http.StatusBadRequest {
+		log.Printf("[WARN] Scaling policy has active workloads (%s) and can't be deleted, removing from state", d.Id())
+		return nil
+	}
```

Comment:
I don't get this part, if we fail to delete it, we remove it from state?

Reply:
I'm pretty sure we can get BadRequest not only for this err when deleting as well. Deleting readonly policy, deleting default policies 🤔.

Reply:
I added it to avoid unremovable state when eg you onboard and offboard your cluster, you would have to go to ui and unnassign all workloads just to remove from state which is annoying 🤔

Reply:
same with default, it you import those you are stuck with them forever as you can't delete it so the only thing is to remove from state and let them be 🤔

Reply:
maybe for now I will remove that check for active workloads and add it if we get any complaints 🤔